contrib/drivers/netapp/ontap/ontap.go

```diff
@@ -0,0 +1,463 @@
+}
+
+func getSnapshotName(id string) string {
+	r := strings.NewReplacer("-", "")
+	return snapshotPrefix + r.Replace(id)
+}
+
+// Get LUN Serial Number
+func (d *Driver) getLunSerialNumber(lunPath string) (string, error) {
+
+	lunSrNumber, err := d.sanStorageDriver.API.LunGetSerialNumber(lunPath)
+	if err != nil {
+		return "", fmt.Errorf("problem reading maps for LUN %s: %v", lunPath, err)
+	}
+
+	return naaPrefix + hex.EncodeToString([]byte(lunSrNumber.Result.SerialNumber())), nil
+}
+
+func (d *Driver) GetVolumeConfig(name string, size int64) (volConfig *storage.VolumeConfig) {
+	volConfig = &storage.VolumeConfig{
+		Version:      VolumeVersion,
+		Name:         name,
+		InternalName: name,
+		Size:         strconv.FormatInt(size*bytesGB, 10),
+		Protocol:     d.sanStorageDriver.GetProtocol(),
+		AccessMode:   accessMode,
+		VolumeMode:   volumeMode,
+		AccessInfo:   utils.VolumeAccessInfo{},
+	}
+	return volConfig
+}
+
+func (d *Driver) GetSnapshotConfig(snapName string, volName string) (snapConfig *storage.SnapshotConfig) {
+	snapConfig = &storage.SnapshotConfig{
+		Version:            SnapshotVersion,
+		Name:               snapName,
+		InternalName:       snapName,
+		VolumeName:         volName,
+		VolumeInternalName: volName,
+	}
+	return snapConfig
+}
+
+func (d *Driver) Setup() error {
+	// Read NetApp ONTAP config file
+	d.conf = &ONTAPConfig{}
+
+	p := config.CONF.OsdsDock.Backends.NetappOntapSan.ConfigPath
+	if "" == p {
+		p = defaultConfPath
+	}
+	if _, err := Parse(d.conf, p); err != nil {
+		return err
+	}
+
+	defer func() {
+		if r := recover(); r != nil {
+			log.Errorf("Unable to instantiate backend")
+		}
+	}()
+
+	empty := ""
+	config := &drivers.OntapStorageDriverConfig{
+		CommonStorageDriverConfig: &drivers.CommonStorageDriverConfig{
+			Version:           d.conf.Version,
+			StorageDriverName: StorageDriverName,
+			StoragePrefixRaw:  json.RawMessage("{}"),
+			StoragePrefix:     &empty,
+		},
+		ManagementLIF: d.conf.ManagementLIF,
+		DataLIF:       d.conf.DataLIF,
+		IgroupName:    d.conf.IgroupName,
+		SVM:           d.conf.Svm,
+		Username:      d.conf.Username,
+		Password:      d.conf.Password,
+	}
+	marshaledJSON, err := json.Marshal(config)
+	if err != nil {
+		log.Fatal("Unable to marshal ONTAP config:  ", err)
+	}
+	configJSON := string(marshaledJSON)
+
+	// Convert config (JSON or YAML) to JSON
+	configJSONBytes, err := yaml.YAMLToJSON([]byte(configJSON))
+	if err != nil {
+		err = fmt.Errorf("invalid config format: %v", err)
+		return err
+	}
+	configJSON = string(configJSONBytes)
+
+	// Parse the common config struct from JSON
+	commonConfig, err := drivers.ValidateCommonSettings(configJSON)
+	if err != nil {
+		err = fmt.Errorf("input failed validation: %v", err)
+		return err
+	}
+
+	d.sanStorageDriver = &ontap.SANStorageDriver{
+		Config: *config,
+	}
+
+	// Initialize the driver.
+	if err = d.sanStorageDriver.Initialize("csi", configJSON, commonConfig); err != nil {
```

Comment:
why need csi here, can it be changed?

Reply:
Its the driver context supported by trident project.
https://github.com/NetApp/trident/blob/master/config/config.go#L108

Reply:
ok